Codechange: Unify structures with sprite sub-tile bounds and simplify bounding boxes.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Lots of different structs contain variations on sub-tile bounds with different naming, and different ordering. Some contain additional offsets but not all.

Bounding box offsets were added later at some point, to allow the sorted bounding box to occupy a different location from the sprite. This behaves very unintuitively, as the visible extents stay with the original position, instead of following the offset.

Without a standard layout for all this, different formats are used all over the place, often involving numbered array access instead, making deciphering harder.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Unify into a single struct, `SpriteBounds`, that can be inherited and passed directly to `AddSortableSpriteToDraw()`.

At the same time, offsets now work more logically: sub-tile bounds now specify the bounding box, and an offset can be applied to the sprite. This means the size of the bounding box is always as specified, not some thing else.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

It's kinda large and intrusive.

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

rubidium42 commented on src/sprite.h:
```diff
@@ -10,28 +10,33 @@
 #include "table/sprites.h"
 
+struct SpriteBounds {
+	PointXyz<int8_t> origin; ///< Position of orthern corner within tile.
```
```suggestion
	PointXyz<int8_t> origin; ///< Position of northern corner within tile.
```

rubidium42 commented on src/tunnelbridge_cmd.cpp:
```diff
@@ -1364,15 +1392,15 @@ static void DrawTile_TunnelBridge(TileInfo *ti)
-		/* Add helper BB for sprite sorting that separates the tunnel from things beside of it. */
-		AddSortableSpriteToDraw(SPR_EMPTY_BOUNDING_BOX, PAL_NONE, ti->x,              ti->y,              BB_data[6], BB_data[7], TILE_HEIGHT, ti->z);
-		AddSortableSpriteToDraw(SPR_EMPTY_BOUNDING_BOX, PAL_NONE, ti->x + BB_data[4], ti->y + BB_data[5], BB_data[6], BB_data[7], TILE_HEIGHT, ti->z);
+		// /* Add helper BB for sprite sorting that separates the tunnel from things beside of it. */
```
Commenting a comment?
```suggestion
		/* Add helper BB for sprite sorting that separates the tunnel from things beside of it. */
```

PeterN committed:
Codechange: Unify structures with sprite sub-tile bounds and simplify bounding boxes.

Lots of different structs contain variations on sub-tile bounds with different naming. Unify into a single struct that can be inherited and passed directly to AddSortableSpriteToDraw.

At the same time, offsets now work more logically: sub-tile bounds now specify the bounding box, and an offset can be applied to the sprite.
